host/xtest/xtest_20000.c

```diff
@@ -0,0 +1,396 @@
+'T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', 'n', ' ',
+'O', 'P', '-', 'T', 'E', 'E', ' ',
+'s', 't', 'o', 'r', 'a', 'g', 'e', ' ', 't', 'e', 's', 't', '.'
+};
+
+static TEEC_Result obj_open(
+	TEEC_Session *sess, void *id, uint32_t id_size, uint32_t flags,
+	uint32_t *obj)
+{
+	TEEC_Operation op;
+	TEEC_Result res;
+	uint32_t org;
+
+	memset(&op, 0, sizeof(op));
+	op.params[0].tmpref.buffer = id;
+	op.params[0].tmpref.size = id_size;
+	op.params[1].value.a = flags;
+	op.params[1].value.b = 0;
+
+	op.paramTypes = TEEC_PARAM_TYPES(
+		TEEC_MEMREF_TEMP_INPUT, TEEC_VALUE_INOUT, TEEC_NONE, TEEC_NONE);
+
+	res = TEEC_InvokeCommand(sess, TA_STORAGE_CMD_OPEN, &op, &org);
+
+	if (res == TEEC_SUCCESS)
+		*obj = op.params[1].value.b;
+
+	return res;
+}
+
+static TEEC_Result obj_create(
+	TEEC_Session *sess, void *id, uint32_t id_size, uint32_t flags,
+	uint32_t attr, void *data, uint32_t data_size, uint32_t *obj)
+{
+	TEEC_Operation op;
+	TEEC_Result res;
+	uint32_t org;
+
+	memset(&op, 0, sizeof(op));
+	op.params[0].tmpref.buffer = id;
+	op.params[0].tmpref.size = id_size;
+	op.params[1].value.a = flags;
+	op.params[1].value.b = 0;
+	op.params[2].value.a = attr;
+	op.params[2].value.b = 0;
+	op.params[3].tmpref.buffer = data;
+	op.params[3].tmpref.size = data_size;
+
+	op.paramTypes = TEEC_PARAM_TYPES(
+		TEEC_MEMREF_TEMP_INPUT, TEEC_VALUE_INOUT, TEEC_VALUE_INPUT,
+		TEEC_MEMREF_TEMP_INPUT);
+
+	res = TEEC_InvokeCommand(sess, TA_STORAGE_CMD_CREATE, &op, &org);
+
+	if (res == TEEC_SUCCESS)
+		*obj = op.params[1].value.b;
+
+	return res;
+}
+
+static TEEC_Result obj_close(
+	TEEC_Session *sess, uint32_t obj)
+{
+	TEEC_Operation op;
+	uint32_t org;
+
+	memset(&op, 0, sizeof(op));
+	op.params[0].value.a = obj;
+
+	op.paramTypes = TEEC_PARAM_TYPES(
+		TEEC_VALUE_INPUT, TEEC_NONE, TEEC_NONE, TEEC_NONE);
+
+	return TEEC_InvokeCommand(sess, TA_STORAGE_CMD_CLOSE, &op, &org);
+}
+
+static int get_ta_dirname(TEEC_UUID *p_uuid, char *buffer)
+{
+
+	if (p_uuid == NULL || buffer == NULL)
+		return 0;
+	
+	return sprintf(buffer, "%08X%04X%04X%02X%02X%02X%02X%02X%02X%02X%02X",
+		       SWAP_BYTES_32(p_uuid->timeLow),
+		       SWAP_BYTES_16(p_uuid->timeMid),
+		       SWAP_BYTES_16(p_uuid->timeHiAndVersion),
+		       p_uuid->clockSeqAndNode[0],
+	               p_uuid->clockSeqAndNode[1],
+		       p_uuid->clockSeqAndNode[2],
+		       p_uuid->clockSeqAndNode[3],
+		       p_uuid->clockSeqAndNode[4],
+		       p_uuid->clockSeqAndNode[5],
+		       p_uuid->clockSeqAndNode[6],
+		       p_uuid->clockSeqAndNode[7]);
+}
+
+static int get_obj_filename(void *file_id, uint32_t file_id_length, char *buffer)
+{
+	char *p = buffer;
+	uint32_t i;
+
+	if (file_id == NULL || buffer == NULL)
+		return 0;
+
+	for (i=0; i<file_id_length; i++)
+		p += sprintf(p, "%02X", ((uint8_t *)file_id)[i]);
+
+	return p-buffer;
+}
+
+static int dump_file(char *ta_dirname, char *obj_filename)
+{
+	char cmd[150];
+
+	sprintf(cmd, "hexdump -C /data/tee/%s/%s", ta_dirname, obj_filename);
```

Comment:
I'm afraid this is a bit fragile (what happens if `hexdump` is not available or not in `$PATH`?).